### DISCUSSION TITLE
Is it possible make importable cmake utility modules(ports) using vcpkg?

### DISCUSSION THREAD

**Discussion by msclock** [microsoft/vcpkg#33331] [Q&A] [answered]
Hallo, i look for a way to make utility ports reduce the painful process on rewriting common modules such as valgrind, sanitizers and others.And if it can be implemented, can anyone give me a simple idea? thanks.

Is the implementation similar to make normal ports?

**Comment by Neumann-A** [comment]
Yes that is possible. Just use `find_package(YourModule CONFIG)` instead of `include(YourModule)`. Just need to put a make config in e.g. `${CURRENT_PACKAGES_DIR}/share/YourModule/yourmodule-config.cmake`. 

(There is a lot of possibilities how to do it)

> Is the implementation similar to make normal pors?

You download whatever and simply directly install into e.g. `${CURRENT_PACKAGES_DIR}/share/YourModule/cmake` you dont need explicit calls to any buildsystem in this case.

**Reply by msclock** [reply]
Thanks for replying, i will try to transform my common utilities using your mentioned way.

**Reply by dg0yt** [reply]
Nitpick: You should actually install to `${CURRENT_PACKAGES_DIR}/share/yourmodule` for a port named `yourmodule`. (Port names must be lower case, and so there will always be a directory `${CURRENT_PACKAGES_DIR}/share/yourmodule`. CMake allows lower-case for the directory name.)

If one cmake file accesses additional scripts or resources, remember to write it relocatable (relative paths, at least based on `${CMAKE_CURRENT_LIST_DIR}` at script runtime).

**Reply by msclock** [reply]
> Nitpick: You should actually install to `${CURRENT_PACKAGES_DIR}/share/yourmodule` for a port named `yourmodule`. (Port names must be lower case, and so there will always be a directory `${CURRENT_PACKAGES_DIR}/share/yourmodule`. CMake allows lower-case for the directory name.)
> 
> If one cmake file accesses additional scripts or resources, remember to write it relocatable (relative paths, at least based on `${CMAKE_CURRENT_LIST_DIR}` at script runtime).

@dg0yt Thanks for your advice! ! The initial idea was to wrap up some of the previous project configurations including valgrind, sanitizer, Sphinx, and so on, making it easy to install references directly from port. This approach is more convenient for project configuration development than the basic modules that originally existed in CMAKE itself.

**Comment by msclock** [comment, marked as answer]
After some practice, I found a way to only share the cmake module. The method is to use vcpkg-cmake-wrapper.cmake to append the module path that needs to be exported to CMAKE_MODULE_PATH  to export the module. Here is a feasible [exmaple](https://github.com/msclock/cmake-registry/blob/master/ports/vcpkg-cmake-module/vcpkg_cmake_module.cmake).
